File: examples/python/decode.py
# Example on how to read WARC files.
import subprocess

import message


def main():
    # Launch the warcat program. The options provided will tell it to write
    # JSON as a line to standard out.
    # Ensure you have warcat on the search path or adjust the path as needed.
    with subprocess.Popen(
        [
            "warcat",
            "export",
            "--input=examples/example.warc",
            "--format=jsonl",
        ],
        stdout=subprocess.PIPE,
    ) as process:
        # Decode each message by using our helper module.
        for msg in message.decode(process.stdout):
            if isinstance(msg, message.Header):
                # We decoded the start of the record.
                print(msg.fields)
            elif isinstance(msg, message.BlockChunk):
                # We decoded the body of the record.
                print(len(msg.data))
            elif isinstance(msg, message.BlockEnd):
                # The end of the record was reached.
                print("---")


if __name__ == "__main__":
    main()

File: examples/c-sharp/WarcatExample/Decode.cs
// Example on how to read WARC files.
using System.Diagnostics;
using System.Text.Json;

namespace WarcatExample;

class Decode
{
    public static void Run()
    {
        var options = Message.Options();

        // Launch the warcat program. The options provided will tell it to write
        // JSON as a line to standard out.
        // Ensure you have warcat on the search path or adjust the path as needed.
        using (var process = new Process())
        {
            process.StartInfo.FileName = "warcat";
            process.StartInfo.ArgumentList.Add("export");
            process.StartInfo.ArgumentList.Add("--input=example.warc");
            process.StartInfo.ArgumentList.Add("--format=jsonl");
            process.StartInfo.RedirectStandardOutput = true;
            process.Start();

            while (true)
            {
                var line = process.StandardOutput.ReadLine();

                if (line == null)
                {
                    break;
                }

                // Decode each message
                var message = JsonSerializer.Deserialize<Message>(line, options)!;

                if (message.Header != null)
                {
                    // We decoded the start of the record.
                    foreach (var field in message.Header.Fields)
                    {
                        Console.WriteLine($"{field[0]}:{field[1]}");
                    }
                }
                else if (message.BlockChunk != null)
                {
                    // We decoded the body of the record.
                    Console.WriteLine($"{message.BlockChunk.Data.Length}");
                }
                else if (message.EndOfFile != null)
                {
                    // The end of the record was reached.
                    Console.WriteLine("---");
                }
            }
        }
    }
}
